scene/2d/parallax_2d.cpp

```diff
@@ -0,0 +1,278 @@
+				set_notify_local_transform(true);
+			}
+#endif // TOOLS_ENABLED
+
+			group_name = "__cameras_" + itos(get_viewport_rid().get_id());
+			add_to_group(group_name);
+			_update_repeat();
+			_update_scroll();
+		} break;
+
+		case NOTIFICATION_INTERNAL_PROCESS: {
+			if (Engine::get_singleton()->is_editor_hint() || (!repeat_size.x && !repeat_size.y) || (!autoscroll.x && !autoscroll.y)) {
+				return;
+			}
+
+			autoscroll_offset += autoscroll * get_process_delta_time();
+			autoscroll_offset = autoscroll_offset.posmodv(repeat_size);
+
+			_update_scroll();
+		} break;
+
+#ifdef TOOLS_ENABLED
+		case NOTIFICATION_LOCAL_TRANSFORM_CHANGED: {
+			if (scroll_dirty) {
+				scroll_dirty = false;
+			} else {
+				set_scroll_offset(get_position());
+			}
+		} break;
+#endif // TOOLS_ENABLED
+
+		case NOTIFICATION_EXIT_TREE: {
+			remove_from_group(group_name);
+		} break;
+	}
+}
+
+void Parallax2D::_camera_moved(const Transform2D &p_transform, const Point2 &p_screen_offset, const Point2 &p_adj_screen_pos) {
+	if (ignore_camera_scroll) {
+		return;
+	}
+
+	set_screen_offset(p_adj_screen_pos);
+}
+
+void Parallax2D::_update_scroll() {
+	if (!is_inside_tree()) {
+		return;
+	}
+
+	Point2 scroll_ofs = screen_offset;
+	Size2 vps = get_viewport_rect().size;
+
+#ifdef TOOLS_ENABLED
+	if (Engine::get_singleton()->is_editor_hint()) {
+		vps = Size2(GLOBAL_GET("display/window/size/viewport_width"), GLOBAL_GET("display/window/size/viewport_height"));
+	}
+#endif // TOOLS ENABLED
+
+	if (limit_begin.x <= limit_end.x - vps.x) {
+		scroll_ofs.x = CLAMP(scroll_ofs.x, limit_begin.x, limit_end.x - vps.x);
+	}
+
+	if (limit_begin.y <= limit_end.y - vps.y) {
+		scroll_ofs.y = CLAMP(scroll_ofs.y, limit_begin.y, limit_end.y - vps.y);
+	}
+
+	scroll_ofs *= scroll_scale;
+
+	if (repeat_size.x) {
+		real_t mod = Math::fposmod(scroll_ofs.x - scroll_offset.x - autoscroll_offset.x, repeat_size.x);
+		scroll_ofs.x = screen_offset.x - mod;
+	} else {
+		scroll_ofs.x = screen_offset.x + scroll_offset.x - scroll_ofs.x;
+	}
+
+	if (repeat_size.y) {
+		real_t mod = Math::fposmod(scroll_ofs.y - scroll_offset.y - autoscroll_offset.y, repeat_size.y);
+		scroll_ofs.y = screen_offset.y - mod;
+	} else {
+		scroll_ofs.y = screen_offset.y + scroll_offset.y - scroll_ofs.y;
+	}
+
+	if (!follow_viewport) {
+		scroll_ofs -= screen_offset;
+	}
+
+#ifdef TOOLS_ENABLED
+	scroll_dirty = true;
+#endif // TOOLS_ENABLED
+
+	set_position(scroll_ofs);
+}
+
+void Parallax2D::_update_repeat() {
+	if (!is_inside_tree()) {
+		return;
+	}
+
+	RID ci = get_canvas_item();
+	Point2 repeatScale = repeat_size * get_scale();
+	RenderingServer::get_singleton()->canvas_set_item_repeat(ci, repeatScale, repeat_times);
+}
+
+void Parallax2D::set_scroll_scale(const Size2 &p_scale) {
+	scroll_scale = p_scale;
+}
+
+Size2 Parallax2D::get_scroll_scale() const {
+	return scroll_scale;
+}
+
+void Parallax2D::set_repeat_size(const Size2 &p_repeat_size) {
+	repeat_size = p_repeat_size.max(Vector2(0, 0));
```

Comment:
```suggestion
	if (p_repeat_size == repeat_size) {
		return;
	}

	repeat_size = p_repeat_size.max(Vector2(0, 0));
```
Missing setter guards. You should add it to every method that has immediate side-effect.